Src/BootCamp.Chapter/Credentials.cs

```diff
@@ -1,72 +1,55 @@
+        public readonly string Password;
         private const string Separator = ",";
 
-        public string CredentialsFile { get; } = "credentials.db";
+        public Credentials(string name, string password)
+        {
+            if (!StringOps.IsValid(name) || !StringOps.IsValid(password))
+            {
+                throw new ArgumentException($"{nameof(name)} or {nameof(password)} must not be null or empty");
+            }
 
-        public Credentials()
+            Name = name;
+            Password = password;
+        }
+
+        public override string ToString()
         {
+            return $"{Name},{Password}";
         }
 
-        public Credentials(string credentialsFile)
+        public bool Equals(Credentials other)
         {
-            CredentialsFile = credentialsFile ?? throw new ArgumentNullException(nameof(credentialsFile));
+            return Name == other.Name && Password == other.Password;
         }
 
-        private bool FindUser(User user)
+        public static bool operator ==(Credentials left, Credentials right)
         {
-            string line;
-            bool found = false;
+            return left.Equals(right);
+        }
 
-            StreamReader reader;
-            try
-            {
-                reader = new StreamReader(CredentialsFile);
-                while ((line = reader.ReadLine()) != null)
-                {
-                    var isValid = TryParse(line, out User credentials);
-                    if (isValid && user.Equals(credentials))
-                    {
-                        found = true;
-                        break;
-                    }
-                }
-            }
-            catch (Exception)
-            {
-                throw new InvalidCredentialsDbFile($"There was an error while trying to work with {CredentialsFile}");
-            }
-            reader?.Close();
-            return found;
+        public static bool operator !=(Credentials left, Credentials right)
+        {
+            return !(left == right);
         }
 
-        private bool AddUser(User user)
+        public override int GetHashCode()
         {
-            if (FindUser(user))
-            {
-                throw new UserAllreadyExistsException("User already exists!");
-            }
-            StreamWriter writer;
-            try
-            {
-                writer = new StreamWriter(CredentialsFile, true);
-                writer.WriteLine($"{user.Name},{user.Password}");
-            }
-            catch (Exception ex)
-            {
-                throw new InvalidCredentialsDbFile($"There was an error while trying to work with {CredentialsFile}", ex);
-            }
+            return HashCode.Combine(Name, Password);
+        }
 
-            writer?.Close();
-            return true;
+        public override bool Equals(object obj)
+        {
+            return obj is Credentials user && Name == user.Name && Password == user.Password;
```

Comment:
You can reuse already existing code- the other Equals overload.
```
Credentials cred = obj as Credentials;
return cred != null && cred.Equals(this)
```

Reply:
this should be:
Credentials? cred = obj as Credentials?;
return cred != null && cred.Equals(this);